File: rust/avp/src/pricing.rs
//! Shared model pricing for AVP agents.
//!
//! The agent stamps `avp.cost.source` on each `assistant_message` so consumers
//! can tell a locally-computed cost from a provider-reported one (or unknown).
//! The price table is data: the default ships in `src/data/prices.json`,
//! embedded at build time. Callers may build or override their own `PriceTable`.

use std::collections::HashMap;

use serde::Deserialize;

/// Provenance of a cost number on the wire. Re-export of the generated wire
/// enum so a computed cost can be stamped directly onto `assistant_message`.
pub use crate::trajectory::AssistantMessageDataAvpCostSource as CostSource;

/// Per-1M-token pricing in USD.
#[derive(Debug, Clone, Deserialize)]
pub struct ModelPrice {
    pub input: f64,
    pub output: f64,
    #[serde(default)]
    pub cache_read: f64,
    #[serde(default)]
    pub cache_write: f64,
}

/// A model-name to price mapping. Callers may build or override their own.
pub type PriceTable = HashMap<String, ModelPrice>;

#[derive(Deserialize)]
struct PriceFile {
    models: HashMap<String, ModelPrice>,
}

const PRICES_JSON: &str = include_str!("data/prices.json");

/// Load the bundled default price table. Returns a fresh, owned table the
/// caller can mutate.
pub fn load_default_prices() -> PriceTable {
    let file: PriceFile =
        serde_json::from_str(PRICES_JSON).expect("bundled prices.json is valid JSON");
    file.models
}

/// Resolve a price by the model the agent put on the wire.
///
/// The bundled table is mirrored from models.dev and keyed by its
/// `<provider>/<model>` id. A wire `model` is either already a slug containing a
/// provider (`openai/gpt-4o`, used as-is) or a bare provider-native string
/// (`claude-sonnet-4-6`) that is qualified with `provider` to form the key
/// (`anthropic/claude-sonnet-4-6`). The exact string is tried first so a custom
/// table keyed by bare names still works.
pub fn resolve_price<'a>(
    prices: &'a PriceTable,
    provider: Option<&str>,
    model: &str,
) -> Option<&'a ModelPrice> {
    if let Some(p) = prices.get(model) {
        return Some(p);
    }
    match provider {
        Some(prov) if !model.contains('/') => prices.get(&format!("{prov}/{model}")),
        _ => None,
    }
}

/// Compute billable USD cost from a turn's token counts, returning the cost
/// and its provenance.
///
/// `provider` is the model's provider (e.g. `"anthropic"`), used to resolve a
/// bare wire model against the `<provider>/<model>`-keyed table; pass `None` if
/// the wire model is already provider-qualified or unknown.
///
/// `input_tokens` follows the AVP convention: cache reads are INCLUDED. Cache
/// reads and writes are billed at their own rates; the fresh remainder gets
/// the regular input rate. An unknown model returns `(0.0, Unknown)` so the
/// caller can flag a silent under-count rather than ship a wrong number.
pub fn compute_cost(
    provider: Option<&str>,
    model: &str,
    input_tokens: u64,
    output_tokens: u64,
    cache_read: u64,
    cache_write: u64,
    prices: &PriceTable,
) -> (f64, CostSource) {
    let Some(p) = resolve_price(prices, provider, model) else {
        return (0.0, CostSource::Unknown);
    };
    let fresh = input_tokens
        .saturating_sub(cache_read)
        .saturating_sub(cache_write);
    let cost = fresh as f64 * p.input / 1_000_000.0
        + cache_read as f64 * p.cache_read / 1_000_000.0
        + cache_write as f64 * p.cache_write / 1_000_000.0
        + output_tokens as f64 * p.output / 1_000_000.0;
    (cost, CostSource::Computed)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn default_table_loads() {
        let t = load_default_prices();
        // Mirrored from models.dev, keyed by `<provider>/<model>`.
        let p = t.get("anthropic/claude-opus-4-7").expect("opus 4.7 in default table");
        assert_eq!(p.input, 5.0);
        assert_eq!(p.output, 25.0);
    }

    #[test]
    fn computes_with_cache_split() {
        // fresh = 1000 - 200 - 100 = 700.
        // 700*5 + 200*0.5 + 100*6.25 + 500*25 = 3500 + 100 + 625 + 12500 = 16725 (per 1e6).
        let prices = load_default_prices();
        // Bare wire model is resolved via the provider to `anthropic/claude-opus-4-7`.
        let (cost, src) = compute_cost(Some("anthropic"), "claude-opus-4-7", 1000, 500, 200, 100, &prices);
        assert!((cost - 0.016725).abs() < 1e-9, "got {cost}");
        assert_eq!(src, CostSource::Computed);
    }

    #[test]
    fn slug_model_resolves_without_provider() {
        // A provider-qualified wire slug is used as the key as-is.
        let prices = load_default_prices();
        let (cost, src) = compute_cost(None, "openai/gpt-4o", 1_000_000, 0, 0, 0, &prices);
        assert_eq!(src, CostSource::Computed);
        assert!((cost - 2.5).abs() < 1e-9, "got {cost}");
    }

    #[test]
    fn unknown_model_is_zero_unknown() {
        let prices = load_default_prices();
        let (cost, src) = compute_cost(Some("nope"), "unknown", 1000, 500, 0, 0, &prices);
        assert_eq!(cost, 0.0);
        assert_eq!(src, CostSource::Unknown);
    }
}

File: avp/bindings/python/src/avp/pricing.py
"""Shared pricing for AVP agents.

Why this lives in `avp` core, not per-agent:
- Anthropic models are reachable through nearly every mainstream agent
  SDK: the raw Anthropic SDK, the Claude Agent SDK, LangChain /
  LangGraph, LlamaIndex, Pydantic AI, the OpenAI-compatible surfaces on
  Bedrock and Vertex, plus the long tail of provider-agnostic
  frameworks. Every AVP adapter that wraps one of these SDKs needs the
  same model-price lookup; a per-adapter copy drifts the moment
  Anthropic ships new pricing.
- The price table is data, not policy. Adapters load it at startup,
  users can override via the public `PriceTable` type, and the on-wire
  cost number is tagged with `avp.cost.source` so downstreams can tell
  a locally-computed number from a provider-reported one (or unknown).

Default ships in `avp/data/prices.json`, mirrored from models.dev and
keyed by its `<provider>/<model>` id (synced by `scripts/sync-prices.py`);
`load_default_prices()` reads it fresh on each call. `compute_cost`
normalizes a wire model (plus its provider) to a key here. To override,
pass a `PriceTable` to your driver / translator at construction.

`COST_SOURCE_*` constants name the audit-source values:
  - `computed`: we did the math locally from a price table
  - `reported`: the API/SDK handed us the number directly
  - `unknown`: no price found and no provider report (cost reported as 0.0)

The agent stamps `avp.cost.source` on each `assistant_message` event so
trajectory consumers can filter / weight by provenance, e.g. an audit
pipeline that trusts reported numbers but flags computed numbers from a
stale price table.
"""

from __future__ import annotations

import json
from collections.abc import Mapping
from importlib import resources
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

CostSource = Literal["computed", "reported", "unknown"]

COST_SOURCE_COMPUTED: CostSource = "computed"
COST_SOURCE_REPORTED: CostSource = "reported"
COST_SOURCE_UNKNOWN: CostSource = "unknown"


class ModelPrice(BaseModel):
    """Per-1M-token pricing in USD."""

    model_config = ConfigDict(frozen=True, extra="forbid")

    input: float = Field(ge=0)
    output: float = Field(ge=0)
    cache_read: float = Field(default=0.0, ge=0)
    cache_write: float = Field(default=0.0, ge=0)


PriceTable = Mapping[str, ModelPrice]


def load_default_prices() -> dict[str, ModelPrice]:
    """Load the bundled default price table from `avp/data/prices.json`.

    Reads on each call so users who patch the file in-place get fresh
    numbers without restarting their process. Returns a fresh dict;
    callers can mutate without side effects.
    """
    raw = resources.files("avp.data").joinpath("prices.json").read_text()
    parsed = json.loads(raw)
    return {
        model: ModelPrice.model_validate(spec) for model, spec in parsed.get("models", {}).items()
    }


def resolve_price(prices: PriceTable, model: str, provider: str | None = None) -> ModelPrice | None:
    """Resolve a price by the model the agent put on the wire.

    The bundled table is mirrored from models.dev and keyed by its
    `<provider>/<model>` id. A wire `model` is either a slug already
    containing a provider (`openai/gpt-4o`, used as-is) or a bare
    provider-native string (`claude-sonnet-4-6`) qualified with
    `provider` to form the key (`anthropic/claude-sonnet-4-6`). The
    exact string is tried first so a custom table keyed by bare names
    still works.

    Gateway caveat: the key is the model's origin slug, so when a Commission
    routes a model through a different storefront (`model: "openai/gpt-4o"` with
    `provider.id: "openrouter"`), this returns the model's *list* price, not the
    gateway's actual price (gateways add margin). Treat the result as a
    best-effort estimate and prefer provider-reported cost when available
    (`avp.cost.source = "reported"`).
    """
    p = prices.get(model)
    if p is not None:
        return p
    if provider and "/" not in model:
        return prices.get(f"{provider}/{model}")
    return None


def compute_cost(
    model: str,
    *,
    provider: str | None = None,
    input_tokens: int,
    output_tokens: int,
    cache_read: int,
    cache_write: int,
    prices: PriceTable,
) -> tuple[float, CostSource]:
    """Compute billable USD cost from a turn's token counts.

    Returns `(cost, source)` so callers can stamp `avp.cost.source` on
    the wire alongside `avp.cost_usd`.

    `provider` is the model's provider (e.g. `"anthropic"`), used to
    resolve a bare wire model against the `<provider>/<model>`-keyed
    table; omit it if the wire model is already provider-qualified.

    `input_tokens` here is the AVP convention (cache reads INCLUDED).
    Cache reads / writes are billed at their own per-token rates; the
    fresh portion gets the regular input rate. If the model isn't in
    the table, returns `(0.0, "unknown")` — caller should warn so
    silent under-counts don't ship.
    """
    p = resolve_price(prices, model, provider)
    if p is None:
        return 0.0, COST_SOURCE_UNKNOWN
    fresh = max(0, input_tokens - cache_read - cache_write)
    cost = (
        fresh * p.input / 1_000_000
        + cache_read * p.cache_read / 1_000_000
        + cache_write * p.cache_write / 1_000_000
        + output_tokens * p.output / 1_000_000
    )
    return cost, COST_SOURCE_COMPUTED
